// http-get/lib/http-get.js
// TODO DRY up the code
require('./object.watch.js');
/* core modules */
var fs = require('fs');
var u = require('url');
var p = require('path');
var http = require('http');
var https = require('https');
/* 3rd party optional module */
var gunzipstream;
try {
	gunzipstream = require('gzbz2/gunzipstream.js');
} catch (e) {}
/* reads the package.json information */
var pack = JSON.parse(fs.readFileSync(p.resolve(__dirname + '/../package.json')).toString('utf8'));
/**
 * Patches in the http.head() method
 * @param options
 * @param cb
 * @return req
 */
http.head = function (options, cb) {
	options.method = 'HEAD';
	var req = http.request(options, cb);
	req.end();
	return req;
};
/**
 * Patches in the https.head() method
 * @param options
 * @param cb
 * @return req
 */
https.head = function (options, cb) {
	options.method = 'HEAD';
	var req = https.request(options, cb);
	req.end();
	return req;
};
/**
 * Makes all the HTTP request headers to be specified as lower case
 * @param hdr
 * @returns hdr
 */
var normalizeHeaders = function (hdr) {
	for (var name in hdr) {
		var lowName = name.toLowerCase();
		var val = hdr[name];
		delete(hdr[name]);
		hdr[lowName] = val;
	}
	return hdr;
};
/**
 * Returns the absolute integer value of the input. Avoids the NaN crap.
 * @param value
 * @returns value
 */
var absInt = function (value) {
	return Math.abs(parseInt(value) | 0);
};
/**
 * Processes the input parameters for get(), head()
 * @param options
 * @param reqId
 * @returns object
 */
var processInput = function (options, reqId) {
	if (reqId == undefined) {
		reqId = 0;
	}
	if (options.redirects) {
		options.redirects = absInt(options.redirects);
	}
	if ( ! options.redirects) {
		options.redirects = 10;
	}
	var url = options.url.trim(), client = http;
	if ( ! url.match(/^https?:\/\//i)) {
		// useful against "URLs" like www.example.com/foo.bar
		url = 'http://' + url;
	}
	if ( ! options.headers) {
		options.headers = {};
	}
	var opt = {
		headers: normalizeHeaders(options.headers)
	};
	if ( ! options.nogzip && gunzipstream) {
		opt.headers['accept-encoding'] = 'gzip';
	} else if (opt.headers['accept-encoding']) {
		delete(opt.headers['accept-encoding']);
	}
	if ( ! opt.headers['user-agent']) {
		opt.headers['user-agent'] = 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:15.0) Gecko/20120427 Firefox/15.0a1';
	}
    opt.headers['referrer'] = url;
    if ( ! options.proxy) {
		url = u.parse(url);
		url.search = url.search || '';
		url.hash = url.hash || '';
		url.path = url.pathname + url.search + url.hash;
		if ( ! url.port) {
			if (url.protocol == 'https:') {
				url.port = 443;
			} else {
				url.port = 80;
			}
		}
		opt.host = url.hostname;
		opt.port = url.port;
		opt.path = url.path;
		if (url.protocol == 'https:') {
			client = https;
		}
		if (url.auth) {
			opt.headers.authorization = new Buffer(url.auth).toString('base64');
		}
	} else {
		opt.host = options.proxy.host;
		opt.port = options.proxy.port;
		opt.path = url;
		if (options.proxy.https) {
			client = https;
		}
	}
    if (opt.host == 'www.proxyget.info')
        opt.headers['cookie'] = 's=0e23be9ac098f271eb4557565602e595; __utma=136040240.1776332072.1341029324.1341029324.1341031571.2; __utmc=136040240; __utmz=136040240.1341029324.1.1.utmcsr=proxy.org|utmccn=(referral)|utmcmd=referral|utmcct=/; __atuvc=1%7C26; c[mangafox.me][/][mfsid]=l6d6vi5l3mqcrllhh0g0lmnm31; c[mangafox.me][/][mfvb_sessionhash]=d31ce6b2e7f2982c97f831259b0b25eb; c[mangafox.me][/][mfvb_lastvisit]=1341029327; TejndE37EDj9463=MTM0MTAyOTM0NTg1NDk0NjMyNDM=; MG_9463=16; svspr=http://www.proxyget.info/browse.php?u=http%3A%2F%2Fmangafox.me%2Fmanga%2Fboku_to_majo_no_jikan%2F&b=10&f=norefer; svsds=8; __qca=P0-753453827-1341029351528; c[quantserve.com][/][mc]=4fee7be7-bdc7d-ebccd-d5316; c[scorecardresearch.com][/][UID]=9c32791-92.123.155.56-1341029389; c[scorecardresearch.com][/][UIDR]=1341029389; __utmb=136040240.18.10.1341031571';

    opt.headers['host'] = opt.host;
    //console.log(opt.headers);
	return {
		options: options,
		opt: opt,
		reqId: reqId,
		client: client,
		url: url
	};
};
/**
 * Prepares the redirect target
 * @param url
 * @param location
 * @returns string
 */
var prepareRedirectUrl = function (url, location) {
	var original = u.parse(url);
	var location = u.parse(location);
	if ( ! location.protocol) {
		location.protocol = original.protocol;
	}
	if ( ! location.host) {
		location.host = original.host;
	}
	return u.format(location);
};
/**
 * Sets the http.Agent.defaultMaxSockets property
 * @param value
 */
var setMaxSockets = function (value) {
	value = absInt(value);
	if ( ! value) { // fallback to the default
		value = 5;
	}
	http.Agent.defaultMaxSockets = value;
};
exports.setMaxSockets = setMaxSockets;
/**
 * The http.get() method
 * @param options
 * @param file
 * @param cb
 * @param reqId
 */
var get = function (options, file, cb, reqId) {
	if (typeof file == 'function') {
		cb = file;
		file = null;
		if (options.maxbody) {
			options.maxbody = absInt(options.maxbody);
			if ( ! options.maxbody) {
				cb(new Error('Invalid options.maxbody specification.'));
				return;
			}
		}
	} else {
		file = file.trim();
		file = p.resolve(file);
	}
	if ( ! options.url) {
		cb(new Error('The options object requires an input URL value.'));
		return;
	}
	var input = processInput(options, reqId);
	var opt = input.opt;
	var client = input.client;
	var url = input.url;
	options = input.options;
	reqId = input.reqId;
	var processRequest = function (req) {
		req.on('response', function (res) {
            if (res.statusCode == 304) {
                console.log(req);
                console.log(res);
                process.exit();
            }
            //console.log(res.statusCode);
			switch (res.statusCode) {
				case 200: // success
					if (options.nogzip || ! gunzipstream) {
						if (res.headers['content-encoding'] == 'gzip') {
							var err = new Error('The server sent gzip content without being requested.');
							err.code = res.statusCode;
							err.headers = res.headers;
							cb(err);
							return;
						}
					}
					var aborted = false;
					res.on('close', function () {
						aborted = true;
						if ( ! options.nogzip) {
							// retry the request without gzip
							delete(options.headers['accept-encoding']);
							options.nogzip = true;
							if (file === null) {
								get(options, cb, reqId);
							} else {
								get(options, file, cb, reqId);
							}
						} else {
							var err = new Error('The server prematurely closed the connection.');
							err.code = res.statusCode;
							err.headers = res.headers;
							cb(err);
						}
					});
					if (file) {
						var transfer = {ended: false};
						if (res.headers['content-encoding'] == 'gzip') {
							var ws = fs.createWriteStream(file);
							ws.on('error', function (err) {
								if ( ! aborted) {
									aborted = true;
									req.abort();
									cb(err);
								}
							});
							transfer.watch('ended', function () {
								transfer.unwatch('ended');
								ws.on('open', function (fd) {
									fs.fsync(fd, function () {
										ws.end();
									});
								});
							});
							ws.on('open', function (fd) {
								transfer.watch('ended', function () {
									transfer.unwatch('ended');
									fs.fsync(fd, function () {
										ws.end();
									});
								});
							});
							ws.on('close', function () {
								if ( ! aborted) {
									var ret = {
										code: 200,
										headers: res.headers,
										file: file
									};
									if (reqId > 0) {
										ret.url = options.url;
									}
									cb(null, ret);
								}
							});
							var gunzip = gunzipstream.wrap(res, {encoding: null});
							gunzip.on('error', function (err) {
								if ( ! aborted) {
									aborted = true;
									req.abort();
									delete(options.headers['accept-encoding']);
									options.nogzip = true;
									get(options, file, cb, reqId);
								}
							});
							gunzip.on('data', function (data) {
								try {
									ws.write(data);
								} catch (e) {} // handled by the error listener
							});
							gunzip.on('end', function () {
								transfer.ended = true;
							});
							gunzip.on('close', function () {
								transfer.ended = true;
							});
						} else { // no gzip response
							var ws = fs.createWriteStream(file);
							transfer.watch('ended', function () {
								transfer.unwatch('ended');
								ws.on('open', function (fd) {
									fs.fsync(fd, function () {
										ws.end();
									});
								});
							});
							ws.on('error', function (err) {
								aborted = true;
								req.abort();
								cb(err);
							});
							ws.on('open', function (fd) {
								transfer.watch('ended', function () {
									transfer.unwatch('ended');
									fs.fsync(fd, function () {
										ws.end();
									});
								});
							});
							ws.on('close', function () {
								if ( ! aborted) {
									var ret = {
										code: 200,
										headers: res.headers,
										file: file
									};
									if (reqId > 0) {
										ret.url = options.url;
									}
									cb(null, ret);
								}
							});
							res.on('data', function (data) {
								if ( ! aborted) {
									try {
										ws.write(data);
									} catch (e) {} // handled by the error listener
								}
							});
							res.on('end', function () {
								transfer.ended = true;
							});
						}
					} else { // buffered response
						var buf = '';
						if (res.headers['content-encoding'] == 'gzip') {
							var gunzip = gunzipstream.wrap(res, {encoding: null});
							gunzip.on('error', function (err) {
								if ( ! aborted) {
									aborted = true;
									req.abort();
									delete(options.headers['accept-encoding']);
									options.nogzip = true;
									get(options, cb, reqId);
								}
							});
							gunzip.on('data', function (data) {
								buf += data;
								if (options.maxbody) {
									if (buf.length > options.maxbody) {
										aborted = true;
										req.abort();
										var err = new Error('Large body detected.');
										err.code = 200;
										err.headers = res.headers;
										cb(err);
									}
								}
							});
							gunzip.on('end', function () {
								if ( ! aborted) {
									var ret = {
										code: 200,
										headers: res.headers,
										buffer: buf
									};
									if (reqId > 0) {
										ret.url = options.url;
									}
									cb(null, ret);
								}
							});
							gunzip.on('close', function () {
								if ( ! aborted) {
									var ret = {
										code: 200,
										headers: res.headers,
										buffer: buf
									};
									if (reqId > 0) {
										ret.url = options.url;
									}
									cb(null, ret);
								}
							});
						} else { // no gzip response
							res.on('data', function (data) {
								if ( ! aborted) {
									buf += data;
									if (options.maxbody) {
										if (buf.length > options.maxbody) {
											aborted = true;
											req.abort();
											var err = new Error('Large body detected.');
											err.code = 200;
											err.headers = res.headers;
											cb(err);
										}
									}
								}
							});
							res.on('end', function () {
								if ( ! aborted) {
									var ret = {
										code: 200,
										headers: res.headers,
										buffer: buf
									};
									if (reqId > 0) {
										ret.url = options.url;
									}
									cb(null, ret);
								}
							});
						}
					}
				break;
				case 300: // redirect
				case 301:
				case 302:
				case 303:
				case 305:
				case 307:
					if ( ! res.headers.location) {
						var err = new Error('Redirect response without location header.');
						err.code = res.statusCode;
						err.headers = res.headers;
						cb(err);
					} else {
						options.url = prepareRedirectUrl(url, res.headers.location);
						if (reqId < options.redirects) {
							reqId++;
							if ( ! file) {
								file = cb;
							}
							get(options, file, cb, reqId);
						} else {
							var err = new Error('Redirect loop detected after ' + Number(reqId) + ' requests.');
							err.code = res.statusCode;
							err.headers = res.headers;
							cb(err);
						}
					}
				break;
				case 304:
					var ret = {
						code: 304,
						headers: res.headers
					};
					if (reqId > 0) {
						ret.url = options.url;
					}
					cb(null, ret);
				break;
				default: // error
					var err = new Error('HTTP Error: ' + Number(res.statusCode));
					err.code = res.statusCode;
					err.headers = res.headers;
					cb(err);
				break;
			}
		});
		req.on('error', function (err) {
			cb(err);
		});
		if (req.setTimeout && typeof options.timeout === 'number') {
			req.setTimeout(options.timeout, function () {
				cb(new Error('Timeout while downloading ' + options.url));
				req.abort();
				return;
			});
		};
		req.end();
	};
	processRequest(client.get(opt));
};
exports.get = get;
/**
 * The http.head() method
 * @param options
 * @param cb
 * @param reqId
 */
var head = function (options, cb, reqId) {
	if ( ! options.url) {
		cb(new Error('The options object requires an input URL value.'));
		return;
	}
	var input = processInput(options, reqId);
	var opt = input.opt;
	var client = input.client;
	var url = input.url;
	options = input.options;
	reqId = input.reqId;
	var processRequest = function (req) {
		req.on('response', function (res) {
			switch (res.statusCode) {
				case 200:
				case 304:
					var ret = {
						code: res.statusCode,
						headers: res.headers
					};
					if (reqId > 0) {
						ret.url = options.url;
					}
					cb(null, ret);
				break;
				case 300: // redirect
				case 301:
				case 302:
				case 303:
				case 305:
				case 307:
					if ( ! res.headers.location) {
						var err = new Error('Redirect response without location header.');
						err.code = res.statusCode;
						err.headers = res.headers;
						cb(err);
					} else {
						options.url = prepareRedirectUrl(url, res.headers.location);
						if (reqId < options.redirects) {
							reqId++;
							head(options, cb, reqId);
						} else {
							var err = new Error('Redirect loop detected after ' + Number(reqId) + ' requests.');
							err.code = res.statusCode;
							err.headers = res.headers;
							cb(err);
						}
					}
				break;
				default: // error
					var err = new Error('HTTP Error: ' + Number(res.statusCode));
					err.code = res.statusCode;
					err.headers = res.headers;
					cb(err);
				break;
			}
		});
		req.on('error', function (err) {
			cb(err);
		});
		if (req.setTimeout && typeof options.timeout === 'number') {
			req.setTimeout(options.timeout, function () {
				cb(new Error('Timeout while downloading ' + options.url));
				req.abort();
				return;
			});
		};
		req.end();
	};
	processRequest(client.head(opt));
};
exports.head = head;
